File: sdk/typescript/src/diagnostics/model_ref.ts
export type ModelRefParseErrorCode =
  | "missing_provider_id"
  | "missing_api"
  | "missing_model_id"
  | "missing_separators"
  | "ambiguous_separators"
  | "invalid_provider_id"
  | "invalid_api"
  | "invalid_percent_escape"
  | "invalid_model_id_encoding"
  | "invalid_utf8_model_id";

export class ModelRefParseError extends Error {
  constructor(
    message: string,
    public readonly code: ModelRefParseErrorCode,
  ) {
    super(message);
    this.name = "ModelRefParseError";
  }
}

export type ParsedModelRef = {
  providerId: string;
  api: string;
  modelId: string;
};

export function parseModelRef(modelRef: string): ParsedModelRef {
  const slashIndex = modelRef.indexOf("/");
  if (slashIndex === -1) {
    throw new ModelRefParseError("model_ref is missing '/' and '@' separators", "missing_separators");
  }
  if (slashIndex === 0) {
    throw new ModelRefParseError("model_ref is missing provider_id segment", "missing_provider_id");
  }

  const firstAtIndex = modelRef.indexOf("@");
  if (firstAtIndex !== -1 && firstAtIndex < slashIndex) {
    throw new ModelRefParseError("model_ref contains ambiguous separators in provider_id", "ambiguous_separators");
  }

  const atOffset = modelRef.slice(slashIndex + 1).indexOf("@");
  if (atOffset === -1) {
    throw new ModelRefParseError("model_ref is missing '@' separator", "missing_separators");
  }
  const atIndex = slashIndex + 1 + atOffset;

  const extraSlashInApi = modelRef.slice(slashIndex + 1, atIndex).indexOf("/");
  if (extraSlashInApi !== -1) {
    throw new ModelRefParseError("model_ref contains ambiguous '/' separators", "ambiguous_separators");
  }
  if (modelRef.slice(atIndex + 1).includes("@")) {
    throw new ModelRefParseError("model_ref contains ambiguous '@' separators", "ambiguous_separators");
  }
  if (modelRef.slice(atIndex + 1).includes("/")) {
    throw new ModelRefParseError("model_ref contains ambiguous '/' separators", "ambiguous_separators");
  }

  const providerId = modelRef.slice(0, slashIndex);
  const api = modelRef.slice(slashIndex + 1, atIndex);
  const encodedModelId = modelRef.slice(atIndex + 1);

  validateProviderId(providerId);
  validateApi(api);

  if (encodedModelId.length === 0) {
    throw new ModelRefParseError("model_ref is missing model_id segment", "missing_model_id");
  }

  validateEncodedModelId(encodedModelId);

  try {
    const modelId = decodeURIComponent(encodedModelId);
    return { providerId, api, modelId };
  } catch {
    throw new ModelRefParseError("model_ref model_id is not valid UTF-8 percent encoding", "invalid_utf8_model_id");
  }
}

function validateProviderId(providerId: string): void {
  if (providerId.length === 0) {
    throw new ModelRefParseError("model_ref is missing provider_id segment", "missing_provider_id");
  }
  if (hasForbiddenSegmentChar(providerId)) {
    throw new ModelRefParseError("provider_id contains forbidden characters", "invalid_provider_id");
  }
}

function validateApi(api: string): void {
  if (api.length === 0) {
    throw new ModelRefParseError("model_ref is missing api segment", "missing_api");
  }
  if (hasForbiddenSegmentChar(api)) {
    throw new ModelRefParseError("api contains forbidden characters", "invalid_api");
  }
}

function hasForbiddenSegmentChar(value: string): boolean {
  return value.includes("/") || value.includes("@") || value.includes("%");
}

function validateEncodedModelId(encodedModelId: string): void {
  let index = 0;
  while (index < encodedModelId.length) {
    const char = encodedModelId[index];
    if (char === "%") {
      const hi = encodedModelId[index + 1];
      const lo = encodedModelId[index + 2];
      if (!hi || !lo || !isHex(hi) || !isHex(lo)) {
        throw new ModelRefParseError("model_id has invalid percent escape", "invalid_percent_escape");
      }
      index += 3;
      continue;
    }

    if (char === "/" || char === "@") {
      throw new ModelRefParseError("model_id contains ambiguous separators", "ambiguous_separators");
    }

    if (!isUnreserved(char)) {
      throw new ModelRefParseError("model_id has non-canonical unescaped characters", "invalid_model_id_encoding");
    }

    index += 1;
  }
}

function isHex(char: string): boolean {
  const code = char.charCodeAt(0);
  return (
    (code >= 0x30 && code <= 0x39) ||
    (code >= 0x41 && code <= 0x46) ||
    (code >= 0x61 && code <= 0x66)
  );
}

function isUnreserved(char: string): boolean {
  if (char.length !== 1) return false;
  const code = char.charCodeAt(0);
  return (
    (code >= 0x41 && code <= 0x5a) ||
    (code >= 0x61 && code <= 0x7a) ||
    (code >= 0x30 && code <= 0x39) ||
    char === "-" ||
    char === "." ||
    char === "_" ||
    char === "~"
  );
}

File: sdk/python/src/oap_sdk/_model_ref.py
"""Internal ``model_ref`` parser.

``model_ref`` is an opaque, server-issued handle (spec §3.2). **Application
code must never parse or construct one** -- take it from ``models.list`` or
``models.resolve`` and pass it through unchanged.

The SDK itself has to look inside for exactly one reason: the provider
protocol's ``complete_request`` / ``stream_request`` payloads carry a resolved
``model`` object (``ai_types.Model``), not a ``model_ref``. This module
reconstructs that object, and is also used for timeout diagnostics. It mirrors
``typescript/src/diagnostics/model_ref.ts``, which is documented the same way.

A ref that does not parse is passed through opaquely rather than rejected, so a
future server-side format change cannot break request submission.
"""

from __future__ import annotations

from typing import NamedTuple
from urllib.parse import unquote

__all__ = ["ParsedModelRef", "ModelRefParseError", "parse_model_ref"]

_UNRESERVED_EXTRA = frozenset("-._~")
_HEX_DIGITS = frozenset("0123456789abcdefABCDEF")


class ParsedModelRef(NamedTuple):
    provider_id: str
    api: str
    model_id: str


class ModelRefParseError(ValueError):
    """``model_ref`` did not match the canonical server format."""

    def __init__(self, message: str, code: str) -> None:
        super().__init__(message)
        self.code = code


def parse_model_ref(model_ref: str) -> ParsedModelRef:
    """Parse ``provider_id/api@percent-encoded-model-id``.

    Raises :class:`ModelRefParseError` when ``model_ref`` is not in the
    canonical form.
    """
    slash_index = model_ref.find("/")
    if slash_index == -1:
        raise ModelRefParseError("model_ref is missing '/' and '@' separators", "missing_separators")
    if slash_index == 0:
        raise ModelRefParseError("model_ref is missing provider_id segment", "missing_provider_id")

    first_at = model_ref.find("@")
    if first_at != -1 and first_at < slash_index:
        raise ModelRefParseError("model_ref contains ambiguous separators in provider_id", "ambiguous_separators")

    at_offset = model_ref.find("@", slash_index + 1)
    if at_offset == -1:
        raise ModelRefParseError("model_ref is missing '@' separator", "missing_separators")

    provider_id = model_ref[:slash_index]
    api = model_ref[slash_index + 1 : at_offset]
    encoded_model_id = model_ref[at_offset + 1 :]

    if "/" in api:
        raise ModelRefParseError("model_ref contains ambiguous '/' separators", "ambiguous_separators")
    if "@" in encoded_model_id or "/" in encoded_model_id:
        raise ModelRefParseError("model_ref contains ambiguous separators", "ambiguous_separators")

    _validate_segment(provider_id, "provider_id")
    _validate_segment(api, "api")

    if not encoded_model_id:
        raise ModelRefParseError("model_ref is missing model_id segment", "missing_model_id")
    _validate_encoded_model_id(encoded_model_id)

    try:
        model_id = unquote(encoded_model_id, errors="strict")
    except UnicodeDecodeError as exc:
        raise ModelRefParseError(
            "model_ref model_id is not valid UTF-8 percent encoding", "invalid_utf8_model_id"
        ) from exc
    return ParsedModelRef(provider_id=provider_id, api=api, model_id=model_id)


def _validate_segment(value: str, name: str) -> None:
    if not value:
        raise ModelRefParseError(f"model_ref is missing {name} segment", f"missing_{name}")
    if any(char in value for char in ("/", "@", "%")):
        raise ModelRefParseError(f"{name} contains forbidden characters", f"invalid_{name}")


def _validate_encoded_model_id(encoded: str) -> None:
    index = 0
    length = len(encoded)
    while index < length:
        char = encoded[index]
        if char == "%":
            if index + 2 >= length or encoded[index + 1] not in _HEX_DIGITS or encoded[index + 2] not in _HEX_DIGITS:
                raise ModelRefParseError("model_id has invalid percent escape", "invalid_percent_escape")
            index += 3
            continue
        if not (char.isascii() and (char.isalnum() or char in _UNRESERVED_EXTRA)):
            raise ModelRefParseError(
                "model_id has non-canonical unescaped characters", "invalid_model_id_encoding"
            )
        index += 1
